ta/crypto_perf/ta_crypto_perf.c

```diff
@@ -465,8 +470,504 @@ TEE_Result cmd_hash_prepare_op(uint32_t param_types, TEE_Param params[4])
+		return TEE_TYPE_DH_KEYPAIR;
+	case ALGO_RSA:
+		return TEE_TYPE_RSA_KEYPAIR;
+	case ALGO_ECDSA:
+		return TEE_TYPE_ECDSA_KEYPAIR;
+	case ALGO_ECDH:
+		return TEE_TYPE_ECDH_KEYPAIR;
+	case ALGO_X25519:
+		return TEE_TYPE_X25519_KEYPAIR;
+	default:
+		EMSG("The algo[%"PRIu32"] is not valid", value);
+		return TEE_TYPE_ILLEGAL_VALUE;
+	}
+}
+
+TEE_Result cmd_asym_prepare_attrs(uint32_t param_types,
+				  TEE_Param params[TEE_NUM_PARAMS])
+{
+	uint32_t exp_param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+						   TEE_PARAM_TYPE_NONE,
+						   TEE_PARAM_TYPE_NONE,
+						   TEE_PARAM_TYPE_NONE);
+
+	if (param_types != exp_param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	crypto_buf = TEE_Malloc(params[0].memref.size, 0);
+	memcpy(crypto_buf, params[0].memref.buffer, params[0].memref.size);
+
+	return unpack_attrs(crypto_buf, params[0].memref.size,
+			    &asym_perf_attrs, &asym_perf_attr_count);
+}
+
+TEE_Result cmd_asym_free_attrs(uint32_t param_types,
+			       TEE_Param params[TEE_NUM_PARAMS])
+{
+	uint32_t exp_param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_NONE,
+						   TEE_PARAM_TYPE_NONE,
+						   TEE_PARAM_TYPE_NONE,
+						   TEE_PARAM_TYPE_NONE);
+
+	if (param_types != exp_param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	TEE_Free(asym_perf_attrs);
+	asym_perf_attrs = NULL;
+	TEE_Free(crypto_buf);
+	crypto_buf = NULL;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result cmd_asym_process_keypair(uint32_t param_types,
+				    TEE_Param params[TEE_NUM_PARAMS])
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	int width_bits = 0;
+	int n = 0;
+	uint32_t exp_param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+						   TEE_PARAM_TYPE_NONE,
+						   TEE_PARAM_TYPE_NONE,
+						   TEE_PARAM_TYPE_NONE);
+
+	if (param_types != exp_param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	width_bits = params[0].value.a;
+	n = params[0].value.b;
+
+	while (n--) {
+		res = TEE_GenerateKey(crypto_obj, width_bits, asym_perf_attrs, asym_perf_attr_count);
+		CHECK(res, "TEE_GenerateKey()", goto out;);
+		TEE_ResetTransientObject(crypto_obj);
+	}
+
+out:
+	return res;
+}
+
+TEE_Result cmd_asym_process_rsa_ecc(uint32_t param_types,
+				    TEE_Param params[TEE_NUM_PARAMS])
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	int n = 0;
+	uint32_t mode = 0;
+	uint32_t exp_param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+						   TEE_PARAM_TYPE_MEMREF_INPUT,
+						   TEE_PARAM_TYPE_MEMREF_OUTPUT,
+						   TEE_PARAM_TYPE_NONE);
+	size_t dummy_size = params[2].memref.size;
+	TEE_Result (*do_asym)(TEE_OperationHandle, const TEE_Attribute *,
+			      uint32_t, const void *, uint32_t, void *,
+			      uint32_t *) = NULL;
+
+	if (param_types != exp_param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	n = params[0].value.a;
+	mode = params[0].value.b;
+
+	if (mode == MODE_ENCRYPT)
+		do_asym = TEE_AsymmetricEncrypt;
+	else if (mode == MODE_DECRYPT)
+		do_asym = TEE_AsymmetricDecrypt;
+	else if (mode == MODE_SIGN)
+		do_asym = TEE_AsymmetricSignDigest;
```

Comment:
Add an `else` to return an error as `do_asym()`as it can be called as `NULL`.
Do the `if (mode == MODE_VERIFY) {...}` first and then the other modes.

Reply:
solved